provider/core/src/refcount.rs

```diff
@@ -0,0 +1,222 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use alloc::boxed::Box;
+use core::fmt::Debug;
+use core::mem::ManuallyDrop;
+use core::ptr::NonNull;
+use yoke::CloneableCart;
+
+#[cfg(not(feature = "sync"))]
+pub(crate) use alloc::rc::Rc as SelectedRc;
+#[cfg(feature = "sync")]
+pub(crate) use alloc::sync::Arc as SelectedRc;
+
+type WrappedType = Box<[u8]>;
+
+const SENTINEL_PTR: *mut WrappedType = core::mem::align_of::<WrappedType>() as *mut WrappedType;
+
+/// A thin wrapper over [`NonNull`] that works with `*const T`
+#[derive(Debug)]
+#[repr(transparent)]
+pub(crate) struct OptionSelectedRcBytes(NonNull<WrappedType>);
+
+impl OptionSelectedRcBytes {
+    #[inline]
+    pub const fn new_sentinel() -> Self {
+        // Safety: `SENTINEL_PTR` is non-null
+        Self(unsafe { NonNull::new_unchecked(SENTINEL_PTR) })
+    }
+
+    #[inline]
+    pub fn from_selected_rc(selected_rc: SelectedRc<WrappedType>) -> Self {
+        let ptr = SelectedRc::into_raw(selected_rc);
+        debug_assert_ne!(
+            ptr, SENTINEL_PTR,
+            "Creating from [A]Rc should never be the sentinel ptr"
+        );
+        debug_assert_ne!(
+            ptr, 0 as *const WrappedType,
+            "Creating from [A]Rc should never be the null ptr"
+        );
+        // Safety: ptr is non-null
+        Self(unsafe { NonNull::new_unchecked(ptr as *mut WrappedType) })
+    }
+
+    #[inline]
+    pub fn take(&mut self) -> Option<SelectedRc<WrappedType>> {
+        let ptr = self.0.as_ptr() as *const WrappedType;
+        // Safety: `SENTINEL_PTR` is non-null
+        self.0 = unsafe { NonNull::new_unchecked(SENTINEL_PTR) };
+        if ptr == SENTINEL_PTR {
+            None
+        } else {
+            // Safety: this type has only 2 constructors, and we already
```

Comment:
nit: this comment should reference the invariants and basically say something along the lines of "according to the invariants it is safe to treat non-sentinel values as an Rc"

though the invariants should also say something about that treatment only being allowed if you're not dropping the Rc